### PR TITLE
Validate API against blacklisted services

### PR DESCRIPTION
**Pull request by strekm**
**Description**

Changes proposed in this pull request:

-Validate API against blacklisted services

**Related issue(s)**See #2694

### CODE REVIEW

File: components/api-controller/pkg/controller/v1alpha2/controller.go [hunk 1]
```diff
@@ -256,6 +265,18 @@ func (c *Controller) validateAPI(newAPI *kymaApi.Api, apiStatusHelper *ApiStatus
 	return setStatus(kymaMeta.Successful)
 }
 
+func (c *Controller) validateVirtualService(newAPI *kymaApi.Api) error {
+	for _, svc := range c.blacklistedServices {
+		separator := "."
+		props := strings.Split(svc, separator)
+
+		if newAPI.Spec.Service.Name == props[0] && newAPI.GetNamespace() == props[1] {
+			return fmt.Errorf("creating VirtualService for %s.%s is forbidden", newAPI.Spec.Service.Name, newAPI.GetNamespace())
```
Comment (Demonsthere) [inline components/api-controller/pkg/controller/v1alpha2/controller.go]: Maybe a better error would be `Service X has been blacklisted by BLACKLISTED_SERVICES env`

File: components/api-controller/pkg/controller/v1alpha2/controller.go [hunk 2]
```diff
@@ -256,6 +265,18 @@ func (c *Controller) validateAPI(newAPI *kymaApi.Api, apiStatusHelper *ApiStatus
 	return setStatus(kymaMeta.Successful)
 }
 
+func (c *Controller) validateVirtualService(newAPI *kymaApi.Api) error {
```
Comment (jakkab) [inline components/api-controller/pkg/controller/v1alpha2/controller.go]: I suggest something like
```
func (c *Controller) validateVirtualService(newAPI *kymaApi.Api) error {

	newApiServiceFullName := fmt.Sprintf("%s.%s", newAPI.Spec.Service.Name, newAPI.GetNamespace())
	
	for _, svc := range c.blacklistedServices {
		
		if svc == newApiServiceFullName {
			return fmt.Errorf("service %s has been blacklisted by BLACKLISTED_SERVICES env", newApiServiceFullName)
		}
	}
	return nil
}
``` 
This way we can avoid multiple conditions and parsing strings.

File: components/api-controller/cmd/controller/main.go [hunk 3]
```diff
@@ -144,6 +146,10 @@ func getCORSConfig() istioNetworkingV1.CORSConfig {
 	}
 }
 
+func readBlacklistedServices(list string) []string {
+	return removeEmptyStrings(strings.Split(list, ","))
+}
+
 func removeEmptyStrings(strings []string) []string {
```
Comment (Tomasz-Smelcerz-SAP) [inline components/api-controller/cmd/controller/main.go]: This function can as well trim whitespaces. It should be trimmed anyway.

File: components/api-controller/cmd/controller/main.go [hunk 4]
```diff
@@ -144,6 +146,10 @@ func getCORSConfig() istioNetworkingV1.CORSConfig {
 	}
 }
 
+func readBlacklistedServices(list string) []string {
+	return removeEmptyStrings(strings.Split(list, ","))
```
Comment (Tomasz-Smelcerz-SAP) [inline components/api-controller/cmd/controller/main.go]: this line: removeEmptyStrings(strings.Split(list, ",")) can be re-used with three other places.

**Pull request merged**